[VPlan] Make CanIV part of region.

fhahn opened the pull request.

fhahn committed:
[VPlan] Make canonical IV part of the region

ayalz reviewed (commented):
Nice! Initial review after taking a first look.

fhahn committed:
[VPlan] Track VPValues instead of VPRecipes in calculateRegisterUsage.

Update calculateRegisterUsageForPlan to track live-ness of VPValues
instead of recipes. This gives slightly more accurate results for
recipes that define multiple values (i.e. VPInterleaveRecipe).

When tracking the live-ness of recipes, all VPValues defined by an
VPInterleaveRecipe are considered alive until the last use of any of
them. When tracking the live-ness of individual VPValues, we can
accurately track the individual values until their last use.

Note the changes in large-loop-rdx.ll and pr47437.ll. This patch
restores the original behavior before introducing VPlan-based liveness
tracking.

fhahn reviewed (commented):
This should be in pretty good shape, still most printing tests needs to be updated.

This now depends on https://github.com/llvm/llvm-project/pull/155301, to be able to track the live-ness of the canonical IV accurartely when estimating register pressure

fhahn committed:
Move prepare transforms.

fhahn committed:
Broadcast writemem

fhahn committed:
Refactor
